shared_client.go

```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dns
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+// SharedClients holds a set of SharedClient instances.
+type SharedClients struct {
+	// SharedClient's lock must not be taken while this lock is held!
+	lock sync.Mutex
+	// clients are created and destroyed on demand, hence 'Mutex' needs to be taken.
+	clients map[string]*SharedClient
+}
+
+func NewSharedClients() *SharedClients {
+	return &SharedClients{
+		clients: make(map[string]*SharedClient),
+	}
+}
+
+// GetSharedClient gets or creates an instance of SharedClient keyed with 'key'.  if 'key' is an
+// empty sting, a new client is always created and it is not actually shared.  The returned 'closer'
+// must be called once the client is no longer needed. Conversely, the returned 'client' must not be
+// used after the closer is called.
+func (s *SharedClients) GetSharedClient(key string, conf *Client, serverAddrStr string) (client *SharedClient, closer func()) {
+	if key == "" {
+		// Simplified case when the client is actually not shared
+		return newSharedClient(conf, serverAddrStr), func() {
+			close(client.requestsToSend)
+			client.conn = nil
+			client.wg.Wait()
+		}
+	}
+	for {
+		// lock for s.clients access
+		s.lock.Lock()
+		// locate client to re-use if possible.
+		client = s.clients[key]
+		if client == nil {
+			client = newSharedClient(conf, serverAddrStr)
+			s.clients[key] = client
+			s.lock.Unlock()
+			// new client, we are done
+			break
+		}
+		s.lock.Unlock()
+
+		// reusing client that may start closing while we wait for its lock
+		client.Lock()
+		if client.refcount > 0 {
+			// not closed, add our refcount
+			client.refcount++
+			client.Unlock()
+			break
+		}
+		// client was closed while we waited for it's lock, discard and try again
+		client.Unlock()
+		client = nil
+	}
+
+	return client, func() {
+		client.Lock()
+		defer client.Unlock()
+		client.refcount--
+		if client.refcount == 0 {
+			// connection close must be completed while holding the client's lock to
+			// avoid a race where a new client dials using the same 5-tuple and gets a
+			// bind error.
+			// The client remains findable so that new users with the same key may wait
+			// for this closing to be done with.
+			close(client.requestsToSend)
+			client.conn = nil
+			client.wg.Wait()
```

Comment:
Same here for swapping the wait and the setting of `conn` to `nil`.

Reply:
Done!, also refactored this repetition to `SharedClient.close()`